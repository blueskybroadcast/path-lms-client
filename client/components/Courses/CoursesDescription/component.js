import React from 'react';
import PropTypes from 'prop-types';
import { Modal } from 'react-bootstrap';

import HtmlEditor from '../../HtmlEditor';

class CoursesDescription extends React.Component {
  static propTypes = {
    description: PropTypes.string,
    isAdmin: PropTypes.bool.isRequired,
    editCourseDescription: PropTypes.func.isRequired
  }

  static defaultProps = {
    description: null
  }

  constructor(props) {
    super(props);
    if (typeof window !== 'undefined') {
      this.ReactQuill = require('react-quill');
    }
  }

  state = {
    show: false,
    description: null
  }

  handleShow = () => this.setState({ show: true })

  handleClose = () => this.setState({ show: false })

  handleDescriptionChange = description => this.setState({ description })

  handleSubmit = (e) => {
    e.preventDefault();
    const { description } = this.state;
    const { editCourseDescription } = this.props;
    editCourseDescription(description).then(() => {
      this.handleClose();
    });
  }

  render() {
    const { description, isAdmin } = this.props;
    const { show } = this.state;

    return (
      <div
        className="description-wrapper"
        id="courses-description"
      >
        <div className="description">
          <div dangerouslySetInnerHTML={{__html: description }} />
          { isAdmin && (
            <>
              <div className="edit-info">
                <a
                  onClick={this.handleShow}
                  data-role="set-description"
                  data-type="courses-description"
                  href="#"
                >
                  Edit Description
                </a>
              </div>

              <Modal
                show={show}
                onHide={this.handleClose}
                size="lg"
                backdrop="static"
              >
                <div className="modal-inner">
                  <section className="modal-form">
                    <header>
                      <h4>Edit Description</h4>
                    </header>
                    <div className="form-wrapper">
                      <form
                        onSubmit={this.handleSubmit}
                        className="simple_form courses_description"
                        acceptCharset="UTF-8"
                        data-remote="true"
                        method="post"
                      >
                        <HtmlEditor
                          defaultValue={description}
                          onChange={this.handleDescriptionChange}
                        />
                        <footer className="confirm">
                          <div className="screen-readers-only">
                            <span className="accessibility-action-info">
                              To Update Description press button with label
                              <strong>&quot;Update Description&quot;</strong>
                              .
                              For cancelling press button with label
                              <strong>&quot;Cancel&quot;</strong>
                              .
                            </span>
                          </div>
                          <button
                            name="button"
                            type="submit"
                            className="btn"
                            data-processable-button="both"
                          >
                            Update Description
                          </button>
                          <button
                            onClick={this.handleClose}
                            className="cancel close-modal"
                            type="button"
                          >
                            Cancel
                          </button>
                        </footer>
                      </form>
                    </div>
                  </section>
                </div>
              </Modal>
            </>
          )}
        </div>
      </div>
    );
  }
}

export default CoursesDescription;
