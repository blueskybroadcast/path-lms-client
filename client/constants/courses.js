export const COURSE_PRICE_PATTERN = new RegExp('\d+(\.\d{0,2})?');
