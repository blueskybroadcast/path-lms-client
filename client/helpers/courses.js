export const convertGroupsToSellableItems = (groupsIds, groupsData) => {
  const sellableItems = {};
  groupsIds.forEach((id, index) => {
    const group = groupsData[id];
    sellableItems[index] = {
      sellable: true,
      groupId: group.id,
      groupName: group.name,
      labelEnabled: false,
      pricesAttributes: {
        0: {
          currency: 'usd',
          price: '',
          label: ''
        }
      }
    };
  });
  return sellableItems;
};

export const updateCoursesWithIcons = (courses, icons) => {
  const coursesWithIcons = { ...courses };

  icons.forEach((icon) => {
    const parsedContent = icon[0]
      .replace('", ', ' ')
      .replace(', "', ' ')
      .replace('"]', '')
      .split(' ');
    // need to replace it with regex
    const id = parsedContent[1];
    const name = parsedContent[2].toLowerCase();
    const count = icon[1];

    if (coursesWithIcons[id].permittedContent) {
      coursesWithIcons[id].permittedContent[name] = count;
    } else {
      coursesWithIcons[id].permittedContent = {
        [name]: count
      };
    }
  });

  return coursesWithIcons;
};
