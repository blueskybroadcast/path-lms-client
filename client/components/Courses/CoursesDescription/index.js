import CoursesDescription from './container';

export default CoursesDescription;
