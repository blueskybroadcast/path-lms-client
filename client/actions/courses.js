import normalize from 'jsonapi-normalizer';

import { FETCH_COURSES } from './types';

import { convertBoolToNumber, formatDate } from '../helpers';
import { updateCoursesWithIcons } from '../helpers/courses';

export const fetchCoursesUserInfo = ids => async (dispatch, getState, api) => {
  const formattedParams = ids.map(id => ({
    id: [id],
    type: 'Course'
  }));
  const res = await api.post('/user_info/load', { icons: formattedParams });
  return res.data.icons;
};

export const fetchCourses = params => async (dispatch, getState, api) => {
  await api.get('/courses', { params })
    .then(({ data }) => {
      const courses = normalize(data);
      const { meta } = data;

      const coursesWithIcons = updateCoursesWithIcons(
        courses.entities.course,
        Object.entries(meta.permittedContentCount)
      );

      courses.entities.category = coursesWithIcons;

      dispatch(fetchCoursesUserInfo(courses.result.course)).then((icons) => {
        icons.forEach((icon) => {
          courses.entities.category[icon.id].statuses = {
            purchase: icon.statuses.purchase
          };
        });

        dispatch({
          type: FETCH_COURSES,
          payload: {
            courses,
            accountBanner: meta.accountBanner,
            coursesDescriptionText: meta.coursesDescriptionText
          }
        });
      });
    });
};

export const addCourse = data => async (dispatch, getState, api) => {
  let sellableItems;
  if (data.selectedSellableItems) {
    sellableItems = {
      example_index: {
        sellable: '0',
        group_id: '',
        group_name: '',
        prices_attributes: {
          0: {
            currency: 'usd'
          }
        }
      }
    };
    Object.values(data.selectedSellableItems).forEach((item, index) => {
      sellableItems[index] = {
        sellable: convertBoolToNumber(item.sellable),
        group_id: item.groupId,
        group_name: item.groupName,
        price_attributes: item.pricesAttributes
      };
    });
  }

  await api.post('/courses', {
    course: {
      active: convertBoolToNumber(data.active),
      name: data.name,
      description: data.descriprion,
      cover_photo_url: data.coverPhotoUrl,
      cover_description: data.coverDescription,
      category_ids: data.categoryIds,
      tracks_attributes: data.trackAttributes,
      // start_date: ,
      // end_date: ,
      featured: convertBoolToNumber(data.featured),
      fullsize_cover_photo_url: data.fullSizeCoverPhotoUrl,
      show_progress: convertBoolToNumber(data.showProgress),
      search_keywords: data.searchKeywords
    },
    item_admins: {
      admin_ids: data.adminIds
    },
    sellable_items: sellableItems,
    // free_cle_settings: {
    //   use_credits: "1",
    //   credits_amount: "15"
    // }
    sellable_item: {
      disable_purchase: convertBoolToNumber(data.limitPurchaseAvailability),
      disable_purchase_after: data.purchaseAvailabilityDate
        ? formatDate(data.purchaseAvailabilityDate)
        : ''
    },
    expiration: {
      expirable: convertBoolToNumber(data.expirable),
      repurchasable: convertBoolToNumber(data.repurchasable),
      expires_at_strategy: convertBoolToNumber(data.expirationByDate),
      expires_at: formatDate(data.expirationDate),
      days_before_expire_strategy: convertBoolToNumber(data.expirationByDays),
      days_before_expire: data.expirationDays
    }
  });

  dispatch(fetchCourses());
};

export const editCourseDescription = text => async (dispatch, getState, api) => {
  await api.post('/account/courses_description', {
    courses_description: {
      text
    }
  });

  dispatch(fetchCourses());
};
