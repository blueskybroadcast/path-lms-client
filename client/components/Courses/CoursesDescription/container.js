import { connect } from 'react-redux';

import CoursesDescription from './component';

import { editCourseDescription } from '../../../actions/courses';

import { coursesDescriptionTextSelector } from '../../../selectors/coursesSelectors';

const mapStateToProps = state => ({
  description: coursesDescriptionTextSelector(state)
});

export default connect(mapStateToProps, { editCourseDescription })(CoursesDescription);
