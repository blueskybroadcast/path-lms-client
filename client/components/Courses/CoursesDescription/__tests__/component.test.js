import React from 'react';
import { shallow } from 'enzyme';
import CoursesDescription from '../component';

import { coursesIds, coursesData } from '../../../../fixtures/courses';

describe('Courses description', () => {
  test('should render CoursesDescription with edit button correctly', () => {
    const wrapper = shallow(
      <CoursesDescription
        description="<ul></ul>"
        isAdmin
        editCourseDescription={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render CoursesDescription without edit button correctly', () => {
    const wrapper = shallow(
      <CoursesDescription
        description="<ul></ul>"
        isAdmin={false}
        editCourseDescription={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
