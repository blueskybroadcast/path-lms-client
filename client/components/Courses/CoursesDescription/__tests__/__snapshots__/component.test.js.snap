// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Courses description should render CoursesDescription with edit button correctly 1`] = `
<div
  className="description-wrapper"
  id="courses-description"
>
  <div
    className="description"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<ul></ul>",
        }
      }
    />
    <div
      className="edit-info"
    >
      <a
        data-role="set-description"
        data-type="courses-description"
        href="#"
        onClick={[Function]}
      >
        Edit Description
      </a>
    </div>
    <ForwardRef(Bootstrap(Modal))
      backdrop="static"
      onHide={[Function]}
      show={false}
      size="lg"
    >
      <div
        className="modal-inner"
      >
        <section
          className="modal-form"
        >
          <header>
            <h4>
              Edit Description
            </h4>
          </header>
          <div
            className="form-wrapper"
          >
            <form
              acceptCharset="UTF-8"
              className="simple_form courses_description"
              data-remote="true"
              method="post"
              onSubmit={[Function]}
            >
              <HtmlEditor
                defaultValue="<ul></ul>"
                onChange={[Function]}
              />
              <footer
                className="confirm"
              >
                <div
                  className="screen-readers-only"
                >
                  <span
                    className="accessibility-action-info"
                  >
                    To Update Description press button with label
                    <strong>
                      "Update Description"
                    </strong>
                    . For cancelling press button with label
                    <strong>
                      "Cancel"
                    </strong>
                    .
                  </span>
                </div>
                <button
                  className="btn"
                  data-processable-button="both"
                  name="button"
                  type="submit"
                >
                  Update Description
                </button>
                <button
                  className="cancel close-modal"
                  onClick={[Function]}
                  type="button"
                >
                  Cancel
                </button>
              </footer>
            </form>
          </div>
        </section>
      </div>
    </ForwardRef(Bootstrap(Modal))>
  </div>
</div>
`;

exports[`Courses description should render CoursesDescription without edit button correctly 1`] = `
<div
  className="description-wrapper"
  id="courses-description"
>
  <div
    className="description"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<ul></ul>",
        }
      }
    />
  </div>
</div>
`;
